File: atcoder/abc309/d.py
from collections import deque

N1, N2, M = map(int, input().split())
adj1 = [[] for _ in range(N1)]
adj2 = [[] for _ in range(N2)]
for _ in range(M):
    u, v = map(int, input().split())
    u, v = u - 1, v - 1
    if u < N1:
        adj1[u].append(v)
        adj1[v].append(u)
    else:
        u, v = u - N1, v - N1
        adj2[u].append(v)
        adj2[v].append(u)


def bfs(adj, root, inf=10**18):
    n = len(adj)
    que = deque()
    dis = [inf for _ in range(n)]

    que.append(root)
    dis[root] = 0

    while len(que) > 0:
        u = que.popleft()
        for v in adj[u]:
            if dis[u] + 1 < dis[v]:
                dis[v] = dis[u] + 1
                que.append(v)

    return max(dis)

d1 = bfs(adj1, 0)
d2 = bfs(adj2, N2 - 1)
print(d1 + d2 + 1)

File: atcoder/abc309/d.rs
#![allow(unused)]

fn main() {
    let inp = readv::<usize>();
    let (n1, n2, m) = (inp[0], inp[1], inp[2]);
    let mut adj1 = vec![vec![]; n1];
    let mut adj2 = vec![vec![]; n2];
    for _ in 0..m {
        let inp = readv::<usize>();
        let (u, v) = (inp[0] - 1, inp[1] - 1);
        if inp[0] < n1 {
            adj1[u].push(v);
            adj1[v].push(u);
        } else {
            adj2[u - n1].push(v - n1);
            adj2[v - n1].push(u - n1);
        }
    }

    let d1 = bfs(&adj1, 0);
    let d2 = bfs(&adj2, n2 - 1);
    println!("{}", d1 + d2 + 1);
}

fn bfs(adj: &Vec<Vec<usize>>, root: usize) -> usize {
    let n = adj.len();
    let inf = !0;
    let mut dis = vec![inf; n];
    let mut que = std::collections::VecDeque::new();
    dis[root] = 0;
    que.push_back(root);
    while let Some(u) = que.pop_front() {
        for &v in adj[u].iter() {
            if dis[v] == inf {
                dis[v] = dis[u] + 1;
                que.push_back(v);
            }
        }
    }
    *dis.iter().max().unwrap()
}

fn read<T: std::str::FromStr>() -> T {
    let mut s = String::new();
    std::io::stdin().read_line(&mut s).ok();
    s.trim().parse().ok().unwrap()
}

fn readv<T: std::str::FromStr>() -> Vec<T> {
    read::<String>()
        .split_ascii_whitespace()
        .map(|t| t.parse().ok().unwrap())
        .collect()
}

fn reads() -> Vec<char> {
    read::<String>().chars().collect::<Vec<char>>()
}

fn join<T: ToString>(arr: &[T], sep: &str) -> String {
    arr.iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join(sep)
}
